Global Styles: Prevent Unwanted ItemGroup List Rendering in Border Panel

im3dabasia opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/67425

## What?
Revised `<ItemGroup>` to use `role="none"` and prevent it from rendering an unnecessary `role="list"` when it is not semantically required.

## Why?

Avoids adding unwanted list roles where a semantic list is not necessary, ensuring accessibility standards are maintained without enforcing unnecessary constraints.

## How?

Updated `<ItemGroup>` to include `role="none"` and an optional as prop for flexibility. Removed instances of `role="list" `where a list structure wasn't required.


## Testing Instructions

1. Add a new post.
2. Add a block, such as a button, that includes shadow support.
3. Inspect the "Drop Shadow" section of the button and observe the rendered ItemsGroup element.
4. Validate that the UI behavior and appearance remain unchanged.

## Screenshots or screencast 


### Before
<img width="1470" alt="Screenshot 2025-01-13 at 3 38 24 PM" src="https://github.com/user-attachments/assets/d13b376d-352d-4769-b464-37aaea2c0c56" />


### After
<img width="1470" alt="Screenshot 2025-01-13 at 3 36 39 PM" src="https://github.com/user-attachments/assets/2fd777f5-62ed-4d04-aa6a-28f85b0c24b2" />

im3dabasia committed:
Global Styles: Fix a11y issue in ItemGroup

afercia commented:
@im3dabasia thanks for your PR.
No objections to this temporary fix but fundamentally it dosn't solve the broader issue. ItemGroup is still used incorrectly and against its documented usage.

Apparently, ItemGroup has been used here only to implement a design for a bordered gray button, which is also inconsistent with the Button component default style variants. To me, this usage is entirely wrong under all aspects, also from a design perspective because it introduces one more button styling that is inconsistent and just overrides the default styles.
If such a button styling is really desired, it should be implemented in the base Button component by adding a new style variant.
Cc @WordPress/gutenberg-components @WordPress/gutenberg-design 

As a quick fix, I have no objections to add `role=none`. I recommend to add a `// TODO: ` inline comment to explain why there is a role=none and to mark this as something that needs a better solution in the future.

afercia commented on packages/block-editor/src/components/global-styles/border-panel.js:
```diff
@@ -298,7 +298,7 @@ export default function BorderPanel( {
 					) : null }
 
-					<ItemGroup isBordered isSeparated>
+					<ItemGroup isBordered isSeparated role="none">
```
I suggest to add an inline comment `// TODO: ` to explain why this role none is used here, as a temporary fix. There are a few examples of todo comments in the codebase for inspiration.

afercia reviewed (dismissed):
See comment.

afercia commented:
> Is there any reason why we couldn't simply refactor it to use a Button component rather than an ItemGroup component?

Ideally, yes. 
It is worth considering the ItemGroup is planned to be refactored so I would think improvements to the component should wait. Hopefully, the new implementation of ItemGroup will enforce correct usage.

To me, one of the main reasons of components being used incorrectly is because design wants a new styling for Buttons or other components. 
- Overriding base components styles and functionality with ad-hoc, local, CSS or code hacks. This should not happen. As mentioned in other issues, it leads to non cohesive UI, maintenance cost in the long term and inconsistencies.
- Base components should be used based on their documented intended usage.
- When a new styling is desired, it should be evaluated and implemented in the base components.
- Alternatively, as being considered for the Button component, the base component should provide a 'naked' version that can be styled ad hoc. This should be handled with care though, to not introduce dozens of ad-hoc styling.
- The fact ItemGroup is used this way just means contributors in this project don't know, don't care, about the actual rendered markup and often use components only based on the desired _visual appearance_ but they ignore semantics, usability and accessibility. I'm surprised these ItemGroup implementations have passed a review and have been merged so lightly without some more thoughtful consideration.

The above isn't to blame individuals. Rather, as I mentioned a few times, I think the process should be improved to prevent such incorrect usage of components. Also, reviews should be more thoughtful.

That said, yes this should be a Button component. I'm not opposed to a temporary fix. A more long term fix should be improving the base components to allow the desired design or, alternatively, reconsider the design to work within the constraints of what the base components provide.

t-hamano commented:
I agree that the current implementation is not ideal, but I don't think it's necessary to solve everything at once.

What is required in this PR? In my opinion, replace it with the `Button` component and solve the accessibility issues.

It will require some overrides with custom styles (gray border), but I think improvements to the `Button` and `ItemGroup` components can be addressed incrementally in follow-up PRs.

afercia commented:
> It will require some overrides with custom styles (gray border)

I think the whole point here is that Design should not implement stylings that aren't provided by the base components. This design trend in this project entirely defeats the purpose of having a library of reusable, consistent, tested, and accessible components. I'd rather change the design here and avoid any custom styles / CSS overrides.

t-hamano commented:
>  I'd rather change the design here and avoid any custom styles / CSS overrides.

Does this mean we pause this PR and add an API to the `Button` component to achieve the grey border?

ciampo commented:
Dropping in with some quick thoughts:

- changing the `Button`'s APIs to add a new gray-bordered style variant just for this usecase doesn't seem like a good idea;
- I don't think that the explanation [here](https://github.com/WordPress/gutenberg/pull/68633#issuecomment-2639212083) is correct. `ItemGroup` was introduced to list items with a given layout / look; these items don't have to be buttons, and even when they effectively are buttons, they don't "compete" with the `Button` element. In this specific scenario, the list is a one-item list. And that one item happens to include a button that triggers a dropdown.
- If the current situation can't stand, and we think the correct implementation is just of have normal `button` semantics, the two least disruptive solutions to me feel like:
  - use `ItemGroup` and override roles, like suggested in this PR;
  - use a vanilla `button` component, and style it as it needs to look (which will avoid overriding `Button` styles)

afercia commented:
> In this specific scenario, the list is a one-item list. And that one item happens to include a button that triggers a dropdown.

@ciampo I think you are missing the point here. 

> the list is a one-item list

It's not a list actually. It's a _broken_ list.

The equivalent in HTML would be as the following examples explained [in the issue](https://github.com/WordPress/gutenberg/issues/67425):

An `<ul>` element that contains a span and no list items:

```
<ul>
    <span>Some text</span>
</ul>
```

A button that contains an `<ul>` element that contains s span:

```
<button>
    <ul> <-- Element ul not allowed as child of element button 
            <span> <-- Element span not allowed as child of element ul 
                Some content
            </span>
    </ul>
</button>
```

Both are not acceptable from a semantic perspective and if this was coded in HTML it would be more evident that it's, I would say, embarrassing and unprofessional HTML. I'm not saying that lightly.

My point is that if the design wants to achieve some custom styling (which is already arguable to me), the implementation should not hack around base components and use them incorrectly. Instead, if these nrew styling are really necessary they should be implemented in the base components as that's the only way to guarantee a consistent UI, maintenance, and avoid bugs. Otherwise, these custom design should simply not be implemented.

ciampo commented:
> It's not a list actually. It's a _broken_ list.

I missed that — and how difficult would it be to add an `<Item>` component as a child of `ItemGroup`, so that we re-instate correct list semantics?

afercia commented:
> how difficult would it be to add an <Item> component as a child of ItemGroup, so that we re-instate correct list semantics?

@ciampo `ItemGroup` already ships with an [Item](https://github.com/WordPress/gutenberg/blob/e757d1d5d8a72a37927a49beea7ebfe72adb65f0/packages/components/src/item-group/item/component.tsx) sub-component that by default has a `role="listitem"` that comes from [useItem](https://github.com/WordPress/gutenberg/blob/e757d1d5d8a72a37927a49beea7ebfe72adb65f0/packages/components/src/item-group/item/hook.ts#L21).

The `ItemGroup` [documentation](https://github.com/WordPress/gutenberg/blob/e757d1d5d8a72a37927a49beea7ebfe72adb65f0/packages/components/src/item-group/item-group/README.md#L11) clearly states that:

> `ItemGroup` should be used in combination with the `Item` sub-component.

However, the component doesn't enforce this and is open to misuse. In fact, contributors to this project did use `ItemGroup` incorrectly a few times for example wrapping it around a Button just to have borders and provide a different styling. I'm surprised that such incorrect usage has been overlooked during code reviews, merged and released so lightly. This kind of process leads to releasing subpar code quality and that's just not OK.

afercia commented:
One more consideration. Not to be picky, But even if we would 'fix' such incorrect usages by adding the missing `Item` so that the HTML would be something like `list > listitem > button`, I would argue that's semantically incorrect as well because a list with just one item is not a _list_ in the first place.

im3dabasia commented:
Hey @afercia, @t-hamano,

What would be the ideal implementation of this PR we are expecting?

Currently, as a temporary fix, I have added the `role="none"` which solves the wrong HTML and also helps improve the a11y. Is this approach OK for this PR?

Let me know what you think.

t-hamano commented:
>   - use `ItemGroup` and override roles, like suggested in this PR;
>   - use a vanilla `button` component, and style it as it needs to look (which will avoid overriding `Button` styles)

It will take more longer time to solve the problems that have been discussed here, so the realistic approach is either of the above for now.

- Personally, `using role=none` feels more hacky than overriding the CSS.
- Other similar UIs use a CSS override approach ([Example](https://github.com/WordPress/gutenberg/blob/3bdadeba1bfe9bf86cbce572299f1655f24265c8/packages/block-editor/src/components/global-styles/style.scss#L62-L65)).

im3dabasia commented:
> * Personally, `using role=none` feels more hacky than overriding the CSS.
> * Other similar UIs use a CSS override approach ([Example](https://github.com/WordPress/gutenberg/blob/3bdadeba1bfe9bf86cbce572299f1655f24265c8/packages/block-editor/src/components/global-styles/style.scss#L62-L65)).

Thank you for the feedback. Since `role=none` seems like a hacky solution, I have removed the ItemGroup and added custom CSS to achieve similar design requirements. I understand that we are gradually moving away from ItemGroup in GB. This solution should help us align with that direction as well

Attaching the screenshot for reference.

<img width="1470" alt="Screenshot 2025-03-25 at 3 12 18 PM" src="https://github.com/user-attachments/assets/0a9a0202-a08f-4c3c-95b3-ab722da9926a" />

im3dabasia committed:
remove: ItemGroup and add Design CSS

im3dabasia commented:
Hey @t-hamano,

I've updated the PR with the suggestions and feedback you provided. This PR now aligns with the approach used in the example you shared earlier. When you have a chance, would you please review it? 

I believe this change is similar to the one in https://github.com/WordPress/gutenberg/pull/68672.

I look forward to your feedback.

t-hamano commented on packages/block-editor/src/components/global-styles/style.scss:
```diff
@@ -27,6 +27,11 @@
 	position: relative;
 }
 
+.block-editor-global-styles__shadow-dropdown {
+	border: 1px solid $gray-300;
+	border-radius: $radius-small;
+}
+
```
I think we can combine several CSS selectors into one, for example:

```scss
.block-editor-global-styles-filters-panel__dropdown,
.block-editor-global-styles__shadow-dropdown {
	display: block;
	padding: 0;
	position: relative;
	border: 1px solid $gray-300;
	border-radius: $radius-small;
}
```

im3dabasia committed:
chore: Combine CSS

im3dabasia commented:
I've made the requested changes. 

Since you've reviewed this PR, I wanted to mention I solved a similar issue here: https://github.com/WordPress/gutenberg/pull/68631. Would appreciate your thoughts on that approach as well.

t-hamano reviewed (approved):
Thanks for the update!

Finally, we should be able to delete this duplicate code: https://github.com/WordPress/gutenberg/pull/68633/files#diff-bd8c41ed80c9a311a6260e15c83b21bdd44de2db3a7c42d76324ea45395ad281R64-R67

im3dabasia committed:
chore: Remove additional CSS
